### PR TITLE
Feature/snapshot weights

### PR DESCRIPTION
**Pull request by martinhjel**
As discussed in #639 

## Changes proposed in this Pull Request
- Add flag to compute snapshot_weightings with timedelta if snapshots are of type pd.DatetimeIndex. Default is False.
- Add argument for selecting snapshot weightings, set default to 1.0.
- Add pandapower and scikit-learn to environment.yaml. Limit pandas version as pandapower fails for version >=2.0 of pandas due to .iteritems(). 
- Add type hints.
- Add pyproject.toml file for project configurations to align with [PEP 518](https://peps.python.org/pep-0518/) (should perhaps have been added in a separate PR..)

## Checklist

- [x] Code changes are sufficiently documented; i.e. new functions contain docstrings and further explanations may be given in `doc`.
- [x] Unit tests for new features were added (if applicable).
- [x] Newly introduced dependencies are added to `environment.yaml`, `environment_docs.yaml` and `setup.py` (if applicable).
- [x] A note for the release notes `doc/release_notes.rst` of the upcoming release is included.
- [x] I consent to the release of this PR's code under the MIT license.

### CODE REVIEW

**Comment by FabianHofmann** [comment]
Hey @martinhjel! Thanks for your contribution! 

The PR looks nice  :) But it is mixing topics a bit. I am wondering about the pandas limitation. The pandapower repo also uses pandas >= 2.0 in there CI (see https://github.com/e2nIEE/pandapower/actions/runs/5043342658/jobs/9045012640#step:4:53). Did you update to their latest package?

**Comment by martinhjel** [comment]
Found out I had pandapower 2.8.0 installed, updated to 2.13, and the `AttributeError: 'Series' object has no attribute 'iteritems'` disappeared. Removed the dependency now.

File: doc/release_notes.rst [hunk 1]
```diff
@@ -17,6 +17,10 @@ Upcoming Release
   ``pypsa.clustering.spatial``. The module ``pypsa.networkclustering`` is now
   deprecated but all functionality will continue to be accessible until the next
   major version.
+* The function ``n.set_snapshots()`` now takes two arguments; ``default_snapshot_weightings``
+  that can be used to change the default snapshot weightings, and ``weightings_from_timedelta``
```
Comment (FabianHofmann) [inline doc/release_notes.rst:21]:
```suggestion
  to change the default snapshot weightings, and ``weightings_from_timedelta``
```

File: pypsa/components.py [hunk 2]
```diff
@@ -454,34 +455,64 @@ def meta(self, new):
...
 
         Parameters
         ----------
-        snapshots : list or pandas.Index
+        snapshots : list, pandas.Index, pd.MultiIndex or pd.DatetimeIndex
             All time steps.
+        default_snapshot_weightings: float
+            The default weight for each snapshot. Defaults to 1.0.
+        weightings_from_timedelta: bool
+            Use the timedelta of `snapshots` as `snapshot_weightings` if `snapshots` is of type `pd.DatetimeIndex`.  Defaults to False.
```
Comment (FabianHofmann) [inline pypsa/components.py:480]:
```suggestion
            Wheter to use the timedelta of `snapshots` as `snapshot_weightings` if `snapshots` is of type `pd.DatetimeIndex`.  Defaults to False.
```

**Review by FabianHofmann — CHANGES_REQUESTED** [review]
Just some points regarding the docs

**Comment by martinhjel** [comment]
@FabianHofmann, is there anything else I have to do to complete this PR?

**Review by FabianHofmann — APPROVED** [review]

**Pull request merged**